core/drivers/crypto/hisilicon/hpre_rsa.c

```diff
@@ -0,0 +1,1397 @@
+#include <rng_support.h>
+#include <stdlib_ext.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee/tee_cryp_utl.h>
+#include <trace.h>
+
+#include "hpre_main.h"
+#include "hpre_rsa.h"
+
+static enum hisi_drv_status hpre_rsa_fill_addr_params(struct hpre_rsa_msg *msg,
+						      struct hpre_sqe *sqe)
+{
+	switch (msg->alg_type) {
+	case HPRE_ALG_KG_CRT: /* KEY GEN */
+		sqe->key = msg->in_dma;
+		sqe->out = msg->out_dma;
+		return HISI_QM_DRVCRYPT_NO_ERR;
+	case HPRE_ALG_NC_NCRT:
+	case HPRE_ALG_NC_CRT:
+		if (msg->is_private) {
+			/* DECRYPT */
+			sqe->key = msg->prikey_dma;
+			sqe->in = msg->in_dma;
+			sqe->out = msg->out_dma;
+		} else {
+			/* ENCRYPT */
+			sqe->key = msg->pubkey_dma;
+			sqe->in = msg->in_dma;
+			sqe->out = msg->out_dma;
+		}
+		return HISI_QM_DRVCRYPT_NO_ERR;
+	default:
+		EMSG("Invalid alg_type[%"PRIu32"]", msg->alg_type);
+		return HISI_QM_DRVCRYPT_IN_EPARA;
+	}
+}
+
+static enum hisi_drv_status hpre_rsa_fill_sqe(void *bd, void *info)
+{
+	struct hpre_rsa_msg *msg = info;
+	struct hpre_sqe *sqe = bd;
+
+	sqe->w0 = msg->alg_type | SHIFT_U32(0x1, HPRE_DONE_SHIFT);
+	sqe->task_len1 = TASK_LENGTH(msg->key_bytes);
+
+	return hpre_rsa_fill_addr_params(msg, sqe);
+}
+
+static enum hisi_drv_status hpre_rsa_parse_sqe(void *bd, void *info __unused)
+{
+	struct hpre_sqe *sqe = bd;
+	uint16_t err = 0;
+	uint16_t done = 0;
+
+	err = HPRE_TASK_ETYPE(sqe->w0);
+	done = HPRE_TASK_DONE(sqe->w0);
+	if (done != HPRE_HW_TASK_DONE || err) {
+		EMSG("HPRE do rsa fail! done=0x%"PRIX16", etype=0x%"PRIX16,
+		     done, err);
+		return HISI_QM_DRVCRYPT_IN_EPARA;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static TEE_Result hpre_rsa_do_task(void *msg)
+{
+	struct hisi_qp *rsa_qp = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+
+	rsa_qp = hpre_create_qp(HISI_QM_CHANNEL_TYPE0);
+	if (!rsa_qp) {
+		EMSG("Fail to create rsa qp");
+		return TEE_ERROR_BUSY;
+	}
+
+	rsa_qp->fill_sqe = hpre_rsa_fill_sqe;
+	rsa_qp->parse_sqe = hpre_rsa_parse_sqe;
+	ret = hisi_qp_send(rsa_qp, msg);
+	if (ret) {
+		EMSG("Fail to send task, ret=%d", ret);
+		res = TEE_ERROR_BAD_STATE;
+		goto done_proc;
+	}
+
+	ret = hisi_qp_recv_sync(rsa_qp, msg);
+	if (ret) {
+		EMSG("Recv task error, ret=%d", ret);
+		res = TEE_ERROR_BAD_STATE;
+		goto done_proc;
+	}
+
+done_proc:
+	hisi_qm_release_qp(rsa_qp);
+
+	return res;
+}
+
+static int32_t hpre_rsa_get_random_data(void *prng __unused,
+					uint8_t *buf, size_t len)
+{
+	if (hw_get_random_bytes(buf, len)) {
+		EMSG("Fail to get random data");
+		return MBEDTLS_ERR_MPI_FILE_IO_ERROR;
+	}
+
+	return 0;
+}
+
+static void hpre_rsa_fill_p_q(struct rsa_keypair *key, mbedtls_mpi *p,
+			      mbedtls_mpi *q)
+{
+	crypto_bignum_copy(key->p, (struct bignum *)p);
+	crypto_bignum_copy(key->q, (struct bignum *)q);
+}
+
+static int32_t hpre_rsa_gen_p_q(size_t nbits, struct rsa_keypair *key)
+{
+	int32_t prime_quality = 0;
+	mbedtls_rsa_context rsa = { };
```

Comment:
You should use the crypto API from `<crypto/crypto.h>` as an abstraction instead of reaching directly into mbedtls. `sw_crypto_acipher_gen_rsa_key()` might be useful here.

Reply:
But I just want to generate p and q here, and then generate rsa_keypair from p and q by hardware.
sw_crypto_acipher_gen_rsa_key() might generate the whole rsa_keypair directly.

Reply:
You've already generated a keypair before calling `hpre_rsa_fill_p_q()`, only the CRT parameters remain.
The heavy lifting has already been done, even if you decide to throw away `D` and `E`. What is `hpre_rsa_fill_p_q()` supposed to do? Recreate `D` and `E` and then deduce the CRT parameters?
By the way, `hpre_rsa_gen_p_q()` looks like a slightly hacked-up copy of `mbedtls_rsa_gen_key()`.

Reply:
When we call hpre_rsa_gen_keypiar, we already have key->e.
We hope hpre_rsa_keypair _init only generate key->p and key->q,The function of hpre_rsa_fill_p_q is to copy p and q to rsa_keypair.
And then, hpre_rsa_init and hpre_rsa_do_task use hardware to calculate N,D(ncrt mode) or dP,dQ ,qInv,N,D(crt mode).
The input is key->e,key->p and key->q.

Do you mean I can use mbedtls_rsa_gen_key() instead of hpre_rsa_gen_p_q, and just throw away D and E?

Reply:
> When we call hpre_rsa_gen_keypiar, we already have key->e.

You're right, that's an input parameter.

> We hope hpre_rsa_keypair _init only generate key->p and key->q,The function of hpre_rsa_fill_p_q is to copy p and q to rsa_keypair. And then, hpre_rsa_init and hpre_rsa_do_task use hardware to calculate N,D(ncrt mode) or dP,dQ ,qInv,N,D(crt mode). The input is key->e,key->p and key->q.

The process to generate `P` and `Q` involves a few checks to see that the values are good. This is an iterative process that is restarted if any of the tests fail. Checking that the produced `D` is large enough is one of the checks. So the assumption that you can ask for just two good values for `P` and `Q` leaves a few holes.

What is the purpose of using hardware to calculate `N`, `D`, and the CRT parameters? Is it more secure? Is it faster?

Without trying to benchmark `hpre_rsa_gen_p_q()` I believe that most of the time is spent in the calls to `mbedtls_mpi_gen_prime()` when generating the key.

> 
> Do you mean I can use mbedtls_rsa_gen_key() instead of hpre_rsa_gen_p_q, and just throw away D and E?

No, but if that's the best you can do with the hardware when generating an RSA key why bother with the hardware?

Reply:
So, I should do rsa_gen_keypair just with software, not implement this mode with hardware?

Reply:
I mean I should just not register hpre_rsa_genkeypair or call mbedtls_rsa_gen_key() in hpre_rsa_genkeypair without hardware.

Reply:
> So, I should do rsa_gen_keypair just with software, not implement this mode with hardware?

Yes, we can do that for now. We can improve that later if someone comes up with something better.

The easiest should be to register `.gen_keypair = sw_crypto_acipher_gen_rsa_key`. Please look at the other `sw_crypto_acipher_*()` functions. It looks like for instance `sw_crypto_acipher_alloc_rsa_keypair()` and `sw_crypto_acipher_free_rsa_keypair()` does the same thing as `hpre_rsa_allocate_keypair()` and `hpre_rsa_free_keypair()`.

Reply:
Get it. Thank you very much.

Reply:
solved